File: src/Scripting/Ruby/patterns/distributed_proxy.rb
# frozen_string_literal: true

module RubyPatterns
  module DistributedProxy
    module_function

    def run
      remote = ->(id) { { id: id, name: 'Ada' } }
      proxy = Struct.new(:id, :remote) do
        def name = remote.call(id)[:name]
      end
      raise 'Distributed Proxy failed' unless proxy.new(7, remote).name == 'Ada'
      true
    end
  end
end

RubyPatterns::DistributedProxy.run if $PROGRAM_NAME == __FILE__

File: src/Web/JavaScriptJS/patterns/distributed_proxy.js
const assert = require("node:assert/strict");
/** Hides remote invocation details behind a local-looking proxy contract. */
function run() {
  const transport = { request: (method, id) => ({ method, id, name: "Ada" }) };
  const userProxy = { get: (id) => transport.request("GET /users/:id", id) };
  assert.deepEqual(userProxy.get("42"), { method: "GET /users/:id", id: "42", name: "Ada" });
  return "distributed-proxy";
}
module.exports = { run };
if (require.main === module) run();
